src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+
+    // Filtered and ranked candidate range
+    const auto first = begin(candidate_vias);
+    const auto last = it;
+    const auto number_of_candidate_vias = last - first;
+
+    std::cout << ">>> number of filtered candidates: " << number_of_candidate_vias << std::endl;
+
+    // Reconstruct packed paths from the heaps.
+    // The recursive path unpacking below destructs heaps.
+    // We need to save all packed paths from the heaps upfront.
+
+    const auto extract_packed_path_from_heaps = [&](WeightedViaNode via) {
+        auto packed_path = retrievePackedPathFromHeap(forward_heap, reverse_heap, via.node);
+        return WeightedViaNodePackedPath{std::move(via), std::move(packed_path)};
+    };
+
+    std::vector<WeightedViaNodePackedPath> weighted_packed_paths;
+    weighted_packed_paths.reserve(1 + number_of_candidate_vias);
+
+    // Store shortest path
+    WeightedViaNode shortest_path_weighted_via{shortest_path_via, shortest_path_weight};
+    weighted_packed_paths.push_back(extract_packed_path_from_heaps(shortest_path_weighted_via));
+
+    // Store all alternative packed paths (if there are any).
+    auto into = std::back_inserter(weighted_packed_paths);
+    std::transform(first, last, into, extract_packed_path_from_heaps);
+
+    // Filter packed paths with heuristics
+
+    auto alternative_paths_last = end(weighted_packed_paths);
+
+    alternative_paths_last = filterPackedPathsByLocalOptimality(weighted_packed_paths[0],
+                                                                forward_heap, // paths for s, via
+                                                                reverse_heap, // paths for via, t
+                                                                begin(weighted_packed_paths) + 1,
+                                                                alternative_paths_last);
+
+    const auto number_of_alternative_paths =
+        alternative_paths_last - (begin(weighted_packed_paths) + 1);
+
+    // Todo: refactor - this is ugly af
+
+    auto number_of_filtered_alternative_paths =
+        std::min(static_cast<std::size_t>(kMaxAlternativesToUnpack),
+                 static_cast<std::size_t>(number_of_alternative_paths));
+
+    for (std::size_t i = 1; i < number_of_filtered_alternative_paths; ++i)
+    {
+        for (std::size_t j = 0; j < i; ++j)
+        {
+            alternative_paths_last =
+                filterPackedPathsByCellSharing(weighted_packed_paths[j],
+                                               partition,
+                                               begin(weighted_packed_paths) + i,
+                                               alternative_paths_last);
+        }
+    }
+
+    number_of_filtered_alternative_paths = std::min(
+        static_cast<std::size_t>(kMaxAlternativesToUnpack),
+        static_cast<std::size_t>(alternative_paths_last - (begin(weighted_packed_paths) + 1)));
+
+    // ^ refactor
+
+    const auto paths_first = begin(weighted_packed_paths);
```

Comment:
No `weighted_packed_paths.front()` is the shortest path, pushed in line 600 above